internal/security/fileprovider/tokenconfig.go

```diff
@@ -80,3 +88,77 @@ func LoadTokenConfig(fileOpener fileioperformer.FileIoPerformer, path string, to
 
 	return nil
 }
+
+// GetTokenConfigFromEnv function gets a list of token service keys
+// from environment variable and populates the default configuration with
+// default token parameters and policies
+// the function returns a TokenConfFile map instance and error if any
+// if the environment variable is not present or the value of that is empty
+// then it will return empty map
+// if the value for the list is not well-formed, not comma-separated
+// then it will return an error
+func GetTokenConfigFromEnv() (TokenConfFile, error) {
+	emptyTokenConfig := make(TokenConfFile)
+
+	addTokenList := os.Getenv(addSecretstoreTokensEnvKey)
+	if strings.TrimSpace(addTokenList) == "" {
+		return emptyTokenConfig, nil
+	}
+
+	// the list of service names is comma-separated
+	tokenConfigFromEnv := make(TokenConfFile)
+	serviceNameList := strings.Split(addTokenList, ",")
+	// regex for valid service name as key
+	// the service name eventually becomes part of the URL to Vault's API call
+	// / is not ok and backslack \ is not ok comma , is not ok
+	// and the the length of name is upto 512 characters
+	serviceNameRegx := regexp.MustCompile(`^[^.*\/\\\,]*[^\W]*[\w.]{1,512}$`)
```

Comment:
I suspect this regexp is more complicated than it needs to be.  \W is defined as [0-9A-Za-z_]  (we should also allow hyphens).  Better to list what IS allowed versus what is NOT allowed.  Doing both is very confusing.

Reply:
This is least complicated regex I can think of. and hyphens are allowed according to this regex.  You can use this online tool to verify hyphens working and understand each regex's group is doing. https://regex101.com/   
can list what is allowed

Reply:
What does `^[\w.-]{1,512}$` allow that you don't want?

Reply:
that one only allows `\w` anything starting with word character and upto length 512. it is not sufficient for cases with some other special characters or white spaces like:
```
case::should be ok
this is  a test cases
edgex:core:service
```

Reply:
```
^[\w.-: ]{1,512}$
```

Reply:
false positive on this case: `/ddd` which we want to avoid

Reply:
I am wondering if we should the regex part into a single function so that it can be tested and legal/malformed string can be included for future reference?

Reply:
The regexp site has a bug.  Go to https://play.golang.org/ and run the following:

```
package main

import (
	"fmt"
	"regexp"
)

func main() {
	serviceNameRegx := regexp.MustCompile(`^[\w.-]{1,512}$`)
	ismatch := serviceNameRegx.MatchString("/ddd")
	if ismatch {
		fmt.Println("is a match")
	} else {
		fmt.Println("no match")
	}
}
```

Output: no match

Reply:
ok that is a good catch but it still fails this `core:service` if just using `^[\w.-]{1,512}$`.  So it is not that simple.... see this: https://play.golang.org/p/E_Q69IZJTxc

Reply:
... please give me some time to investigate ...

Reply:
> I am wondering if we should extra the regex part into a single function so that it can be tested and legal/malformed strings can be included for future reference?

@tingyuz  it is kind of covered already in this test function: https://github.com/jim-wang-intel/edgex-go/blob/2266-addtion-vault-tokens/internal/security/fileprovider/provider_test.go#L270 
and we can add more negative or positive test cases to this test function as many as we want.

Reply:
Sounds good - I didn't review the test file before putting the comments. It would be helpful to add more tests to show what a required format would be.

Reply:
Added two more cases just now.

Reply:
Move the hyphen to the end of the regexp

`^[\w.: -]{1,512}$`

What was happening is that `.-:` is being interpreted as a character range.  The hyphen needs to be last.

Reply:
I think the list of allowed symbols or charaters in the regex  becomes unmanageable according to the following rule of allowance:
```go
	// anything containing the followings are NOT ok:
	// slash (/) is not ok [since it is a directory delimiter]
	// backslash (\) is not ok [since it is a directory delimiter]
	// comma (,) is not ok [since it is a separator in the environment list itself]
	// any other characters than the above is ok
	// and the the length of name is upto 512 characters
```

so instead of saying which are allowed, we should do something not allowed and that is original idea I had in mind when doing the regex of that. Not sure how to express this more precisely.

Reply:
Then we agree to disagree.   From a security perspective blacklisting is deeply flawed.  My advice is not to do it.

 https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html

"It is a common mistake to use black list validation in order to try to detect possibly dangerous characters and patterns like the apostrophe ' character, the string 1=1, or the <script> tag, but this is a massively flawed approach as it is trivial for an attacker to bypass such filters."

Reply:
ok, so how about this list? 
```^[\w.: \!\@\#\^\&\?\-=\+;\|\'\"\<\>\{\}\[\]]{1,512}$```
basically it covers most of characters (white listed in that point of view) and note that the slash / backslash \\ and comma, are not in the list...

Reply:
https://tools.ietf.org/html/rfc3986#page-12

The following characters are reserved characters for URI's and should not be allowed

```
      gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"

      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
```

Reply:
ok, sounds good; will construct according to this guideline.